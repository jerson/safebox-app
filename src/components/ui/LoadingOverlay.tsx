import React, { useState, useEffect } from 'react';
import { Animated, StyleSheet, ViewStyle } from 'react-native';
import Loading, { LoadingProps } from './Loading';
import Colors from '../../modules/constants/Colors';

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.grey1,
    position: 'absolute',
    left: 0,
    bottom: 0,
    right: 0,
    top: 0,
    flex: 1,
    zIndex: 1000,
    justifyContent: 'center'
  } as ViewStyle
});

export interface LoadingOverlayProps extends LoadingProps {
  isLoading: boolean;
  duration?: number;
}

function LoadingOverlay(props: LoadingOverlayProps) {
  const { style, isLoading, duration, ...extraProps } = props;

  const [isVisible, setIsVisible] = useState(false);
  const [fadeAnim] = useState(new Animated.Value(0));

  let animation: Animated.CompositeAnimation;

  useEffect(() => {
    isLoading && animatedFade();
  }, []);

  useEffect(() => {
    animatedFade();
  }, [isLoading]);

  const animatedFade = () => {
    setIsVisible(true);

    animation && animation.stop();
    animation = Animated.timing(fadeAnim, {
      toValue: isLoading ? 1 : 0,
      duration
    });

    animation.start(() => {
      setIsVisible(isLoading);
    });
  };

  if (!isVisible) {
    return null;
  }
  return (
    <Animated.View style={[styles.container, { opacity: fadeAnim }, style]}>
      <Loading {...extraProps} />
    </Animated.View>
  );
}
LoadingOverlay.defaultProps = {
  size: 'large',
  color: Colors.primary,
  isLoading: false,
  duration: 100
};
export default LoadingOverlay;
