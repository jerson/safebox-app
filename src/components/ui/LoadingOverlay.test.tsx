import React from 'react';

import * as renderer from 'react-test-renderer';
import LoadingOverlay, { LoadingOverlayProps } from './LoadingOverlay';

const props: LoadingOverlayProps = {
  isLoading: true
};

it('render just views', () => {
  const snapshot = renderer.create(<LoadingOverlay {...props} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
it('render just views !isLoading', () => {
  const snapshot = renderer
    .create(<LoadingOverlay {...props} isLoading={false} />)
    .toJSON();
  expect(snapshot).toMatchSnapshot();
});
