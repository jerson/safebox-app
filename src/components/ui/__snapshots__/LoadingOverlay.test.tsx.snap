// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render just views !isLoading 1`] = `null`;

exports[`render just views 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fafdff",
      "bottom": 0,
      "flex": 1,
      "justifyContent": "center",
      "left": 0,
      "opacity": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 1000,
    }
  }
>
  <View
    style={
      Array [
        false,
        false,
        false,
        undefined,
      ]
    }
  >
    <ActivityIndicator
      animating={true}
      color="#4531cd"
      duration={100}
      hidesWhenStopped={true}
      isLoading={true}
      size="large"
    />
  </View>
</View>
`;
